File: CPE/Exam/150324/UVA-302_John's trip.cpp
#include<bits/stdc++.h>
#define PII pair<int,int>
#define fi first
#define se second
using namespace std;
const int N=2001;
int n,m,u,v,w,s,d[N],ans[N],l;
bool vis[N],e;
vector<PII>G[N];
void dfs(int u){
	for(int i=0;i<G[u].size();i++){
		int v=G[u][i].se,w=G[u][i].fi; 
		if(vis[w]) continue;
        vis[w]=1;
		dfs(v);
        ans[++l]=w;
	}
}
int main(){
    while(cin>>u>>v){
        if(!u&&!v){
            if(!e) break;
            for(int i=1;i<=2000;i++){
                sort(G[i].begin(),G[i].end());
                if(d[i]%2){
                    cout<<"Round trip does not exist.\n";
                    goto T2;
                }
            }
            dfs(s);
            while(l>1) cout<<ans[l--]<<" ";
            cout<<ans[l--]<<endl;
            T2:;
            memset(vis,0,sizeof vis);
            memset(d,0,sizeof d);
            for(int i=1;i<=2000;i++) G[i].clear();
            s=0;
            e=0;
            cout<<endl;
            continue;
        }
        e=1;
        cin>>w;
        if(!s) s=min(u,v);
        G[u].push_back({w,v});
        G[v].push_back({w,u});
        d[u]--;
        d[v]++;
    }
}

File: CPE/Exam/150324/UVA-302_John's trip.py
import sys
sys.setrecursionlimit(1 << 20)
from collections import defaultdict

while True:
    u, v, *rest = map(int, input().strip().split())
    if u == 0 and v == 0:
        break
    edges = [(u, v, rest[0])]
    while True:
        u, v, *rest = map(int, input().strip().split())
        if u == 0 and v == 0:
            break
        edges.append((u, v, rest[0]))

    g = defaultdict(list)
    deg = defaultdict(int)
    for u, v, eid in edges:
        g[u].append((v, eid))
        g[v].append((u, eid))
        deg[u] += 1
        deg[v] += 1

    if any(deg[u] & 1 for u in deg):
        print("Round trip does not exist.")
        print()
        continue

    for u in g:
        g[u].sort(key=lambda x: x[1]) # 按照邊的編號排序
 
    path = []
    used = set()
    def dfs(u):
        while g[u]:
            v, idx = g[u].pop(0)
            if idx in used:
                continue
            used.add(idx)
            dfs(v)
            path.append(idx)

    st = min(edges[0][0], edges[0][1]) # 從第一條邊的較小端點開始
    dfs(st)
    print(*path[::-1])
    print()